File: clients/java/src/main/java/tech/zpkg/client/ZedClient.java
package tech.zpkg.client;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Duration;
import java.util.HexFormat;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.UUID;

/**
 * Java 17+ client for the stable zed-pkg registry lifecycle.
 *
 * <p>The client never follows redirects, never includes the registry bearer on
 * artifact-download requests, bounds response bodies, and verifies artifact
 * SHA-256 values before returning bytes.</p>
 */
public final class ZedClient {
    public static final String DEFAULT_REGISTRY_URL = "https://registry.zpkg.tech";
    public static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(30);
    public static final int MAX_ARTIFACT_BYTES = 100 * 1024 * 1024;

    private static final int DOWNLOAD_SLACK_BYTES = 1024 * 1024;
    private static final int MAX_JSON_BYTES = 2 * 1024 * 1024;
    private static final int MAX_ERROR_BYTES = 16 * 1024;
    private static final int MAX_SEGMENT_BYTES = 256;

    private final URI baseUri;
    private final String token;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    public ZedClient() {
        this(DEFAULT_REGISTRY_URL, null);
    }

    public ZedClient(String registryUrl, String token) {
        this(
                registryUrl,
                token,
                HttpClient.newBuilder()
                        .connectTimeout(DEFAULT_TIMEOUT)
                        .followRedirects(HttpClient.Redirect.NEVER)
                        .build(),
                new ObjectMapper()
        );
    }

    ZedClient(
            String registryUrl,
            String token,
            HttpClient httpClient,
            ObjectMapper objectMapper
    ) {
        this.baseUri = validateBaseUri(registryUrl);
        this.token = normalizeOptional(token);
        this.httpClient = Objects.requireNonNull(httpClient, "httpClient");
        this.objectMapper = Objects.requireNonNull(objectMapper, "objectMapper");
    }

    public PackageMetadata getPackage(String org, String name)
            throws IOException, InterruptedException {
        return requestJson(
                "GET",
                packagePath(org, name),
                false,
                null,
                null,
                PackageMetadata.class
        );
    }

    public VersionMetadata getVersion(String org, String name, String version)
            throws IOException, InterruptedException {
        return requestJson(
                "GET",
                versionPath(org, name, version),
                false,
                null,
                null,
                VersionMetadata.class
        );
    }

    public SearchResponse search(String query) throws IOException, InterruptedException {
        if (query == null) {
            throw new ValidationException("query must not be null");
        }
        return requestJson(
                "GET",
                "/v1/search?q=" + encodeQuery(query),
                false,
                null,
                null,
                SearchResponse.class
        );
    }

    public ClaimOrgResponse claimOrg(String slug) throws IOException, InterruptedException {
        ObjectNode body = objectMapper.createObjectNode().put("slug", requireText(slug, "slug"));
        return requestJson(
                "POST",
                "/v1/orgs",
                true,
                objectMapper.writeValueAsBytes(body),
                "application/json",
                ClaimOrgResponse.class
        );
    }

    public YankResponse setYanked(
            String org,
            String name,
            String version,
            boolean yanked
    ) throws IOException, InterruptedException {
        ObjectNode body = objectMapper.createObjectNode().put("yanked", yanked);
        return requestJson(
                "POST",
                yankPath(org, name, version),
                true,
                objectMapper.writeValueAsBytes(body),
                "application/json",
                YankResponse.class
        );
    }

    public YankResponse yank(String org, String name, String version)
            throws IOException, InterruptedException {
        return setYanked(org, name, version, true);
    }

    public YankResponse restore(String org, String name, String version)
            throws IOException, InterruptedException {
        return setYanked(org, name, version, false);
    }

    /**
     * Download and verify an artifact. The registry bearer is deliberately not
     * attached because {@code download_url} may be a third-party presigned URL.
     */
    public byte[] downloadArtifact(VersionMetadata version)
            throws IOException, InterruptedException {
        Objects.requireNonNull(version, "version");
        URI downloadUri = downloadUri(version);
        int limit = downloadLimit(version.size());

        HttpRequest request = HttpRequest.newBuilder(downloadUri)
                .timeout(DEFAULT_TIMEOUT)
                .header("Accept", "application/octet-stream")
                .GET()
                .build();
        HttpResponse<InputStream> response = httpClient.send(
                request,
                HttpResponse.BodyHandlers.ofInputStream()
        );
        try (InputStream stream = response.body()) {
            if (response.statusCode() < 200 || response.statusCode() >= 300) {
                throw apiException(response.statusCode(), stream);
            }
            String declaredLength = response.headers().firstValue("content-length").orElse(null);
            if (declaredLength != null) {
                try {
                    if (Long.parseLong(declaredLength) > limit) {
                        throw new ArtifactTooLargeException(limit);
                    }
                } catch (NumberFormatException ignored) {
                    // A malformed content-length is not trusted; the streaming bound still applies.
                }
            }
            byte[] artifact = readBounded(stream, limit, true);
            String actual = sha256Hex(artifact);
            if (!actual.equalsIgnoreCase(requireText(version.sha256(), "version.sha256"))) {
                throw new Sha256MismatchException(version.sha256(), actual);
            }
            return artifact;
        }
    }

    /**
     * Publish raw archive bytes as multipart fields {@code meta} and
     * {@code artifact}. The package coordinate is read from
     * {@code meta.manifest.package} and determines the PUT route.
     */
    public PublishResponse publish(JsonNode meta, byte[] artifact)
            throws IOException, InterruptedException {
        Objects.requireNonNull(meta, "meta");
        Objects.requireNonNull(artifact, "artifact");
        if (!meta.isObject()) {
            throw new ValidationException("publish meta must be a JSON object");
        }
        if (artifact.length > MAX_ARTIFACT_BYTES) {
            throw new ArtifactTooLargeException(MAX_ARTIFACT_BYTES);
        }

        JsonNode pkg = meta.path("manifest").path("package");
        if (!pkg.isObject()) {
            throw new ValidationException("publish meta.manifest.package must be an object");
        }
        String org = requireText(pkg.path("org").asText(null), "meta.manifest.package.org");
        String name = requireText(pkg.path("name").asText(null), "meta.manifest.package.name");
        String version = requireText(
                pkg.path("version").asText(null),
                "meta.manifest.package.version"
        );

        String boundary = "zed-" + UUID.randomUUID();
        byte[] body = multipartBody(boundary, meta, artifact, org, name, version);
        return requestJson(
                "PUT",
                versionPath(org, name, version),
                true,
                body,
                "multipart/form-data; boundary=" + boundary,
                PublishResponse.class
        );
    }

    @Override
    public String toString() {
        return "ZedClient(baseUri=" + baseUri + ", token=[REDACTED])";
    }

    URI baseUri() {
        return baseUri;
    }

    private <T> T requestJson(
            String method,
            String path,
            boolean authenticated,
            byte[] body,
            String contentType,
            Class<T> type
    ) throws IOException, InterruptedException {
        HttpResponse<InputStream> response = send(method, path, authenticated, body, contentType);
        try (InputStream stream = response.body()) {
            if (response.statusCode() < 200 || response.statusCode() >= 300) {
                throw apiException(response.statusCode(), stream);
            }
            byte[] document = readBounded(stream, MAX_JSON_BYTES, true);
            try {
                return objectMapper.readValue(document, type);
            } catch (IOException error) {
                throw new ProtocolException("registry returned invalid JSON", error);
            }
        }
    }

    private HttpResponse<InputStream> send(
            String method,
            String path,
            boolean authenticated,
            byte[] body,
            String contentType
    ) throws IOException, InterruptedException {
        HttpRequest.Builder builder = HttpRequest.newBuilder(resolve(path))
                .timeout(DEFAULT_TIMEOUT)
                .header("Accept", "application/json");
        if (authenticated) {
            builder.header("Authorization", "Bearer " + requireToken());
        }
        if (body == null) {
            builder.method(method, HttpRequest.BodyPublishers.noBody());
        } else {
            builder.header("Content-Type", Objects.requireNonNull(contentType, "contentType"))
                    .method(method, HttpRequest.BodyPublishers.ofByteArray(body));
        }
        return httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofInputStream());
    }

    private ApiException apiException(int status, InputStream stream) throws IOException {
        byte[] bounded = readBounded(stream, MAX_ERROR_BYTES, false);
        String responseBody = new String(bounded, StandardCharsets.UTF_8)
                .replace('\r', ' ')
                .replace('\n', ' ');
        String code = "http_" + status;
        try {
            JsonNode value = objectMapper.readTree(bounded);
            if (value != null && value.path("code").isTextual()) {
                code = value.path("code").asText();
            }
        } catch (IOException ignored) {
            // Non-JSON errors retain the stable HTTP-derived code.
        }
        return new ApiException(status, code, responseBody);
    }

    private URI resolve(String path) {
        return URI.create(baseUri.toString() + path);
    }

    private URI downloadUri(VersionMetadata version) {
        String raw = normalizeOptional(version.downloadUrl());
        if (raw == null || !raw.contains("://")) {
            return resolve(artifactPath(version.sha256()));
        }
        final URI uri;
        try {
            uri = new URI(raw);
        } catch (URISyntaxException error) {
            throw new ValidationException("download_url is invalid", error);
        }
        String scheme = uri.getScheme() == null ? "" : uri.getScheme().toLowerCase(Locale.ROOT);
        String host = uri.getHost() == null ? "" : uri.getHost().toLowerCase(Locale.ROOT);
        boolean loopback = host.equals("localhost") || host.equals("127.0.0.1") || host.equals("::1");
        boolean allowed = scheme.equals("https")
                || (scheme.equals("http")
                && (loopback || baseUri.getScheme().equalsIgnoreCase("http")));
        if (!allowed
                || host.isBlank()
                || uri.getRawUserInfo() != null
                || uri.getRawFragment() != null) {
            throw new ValidationException(
                    "download_url must use HTTPS; HTTP is allowed only for loopback or an HTTP development registry"
            );
        }
        return uri;
    }

    private static URI validateBaseUri(String input) {
        if (input == null) {
            throw new ValidationException("registry URL must not be null");
        }
        String trimmed = input.trim();
        while (trimmed.endsWith("/")) {
            trimmed = trimmed.substring(0, trimmed.length() - 1);
        }
        final URI uri;
        try {
            uri = new URI(trimmed);
        } catch (URISyntaxException error) {
            throw new ValidationException("registry URL is invalid", error);
        }
        String scheme = uri.getScheme() == null ? "" : uri.getScheme().toLowerCase(Locale.ROOT);
        if (!(scheme.equals("http") || scheme.equals("https"))
                || uri.getHost() == null
                || uri.getHost().isBlank()
                || uri.getRawUserInfo() != null
                || uri.getRawQuery() != null
                || uri.getRawFragment() != null) {
            throw new ValidationException(
                    "registry URL must be a credential-free absolute HTTP(S) URL without query or fragment"
            );
        }
        return uri;
    }

    private static String packagePath(String org, String name) {
        return "/v1/packages/" + encodeSegment(org, "org") + "/" + encodeSegment(name, "name");
    }

    private static String versionPath(String org, String name, String version) {
        return packagePath(org, name) + "/versions/" + encodeSegment(version, "version");
    }

    private static String yankPath(String org, String name, String version) {
        return versionPath(org, name, version) + "/yank";
    }

    private static String artifactPath(String sha256) {
        return "/v1/artifacts/" + encodeSegment(sha256, "sha256");
    }

    private static String encodeSegment(String value, String name) {
        String checked = requireText(value, name);
        if (checked.getBytes(StandardCharsets.UTF_8).length > MAX_SEGMENT_BYTES
                || checked.indexOf('\0') >= 0) {
            throw new ValidationException(name + " is too large or contains NUL");
        }
        return URLEncoder.encode(checked, StandardCharsets.UTF_8).replace("+", "%20");
    }

    private static String encodeQuery(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20");
    }

    private String requireToken() {
        if (token == null) {
            throw new MissingTokenException();
        }
        return token;
    }

    private static String requireText(String value, String name) {
        if (value == null || value.isBlank()) {
            throw new ValidationException(name + " must not be blank");
        }
        return value;
    }

    private static String normalizeOptional(String value) {
        if (value == null) {
            return null;
        }
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }

    private static int downloadLimit(long declaredSize) {
        if (declaredSize <= 0) {
            return MAX_ARTIFACT_BYTES;
        }
        long withSlack;
        try {
            withSlack = Math.addExact(declaredSize, DOWNLOAD_SLACK_BYTES);
        } catch (ArithmeticException ignored) {
            withSlack = Long.MAX_VALUE;
        }
        return (int) Math.min(withSlack, MAX_ARTIFACT_BYTES);
    }

    private byte[] multipartBody(
            String boundary,
            JsonNode meta,
            byte[] artifact,
            String org,
            String name,
            String version
    ) throws IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream(
                Math.min(artifact.length + 4096, MAX_ARTIFACT_BYTES)
        );
        writeAscii(output, "--" + boundary + "\r\n");
        writeAscii(output, "Content-Disposition: form-data; name=\"meta\"\r\n");
        writeAscii(output, "Content-Type: application/json\r\n\r\n");
        output.write(objectMapper.writeValueAsBytes(meta));
        writeAscii(output, "\r\n--" + boundary + "\r\n");
        String filename = safeFilename(org + "-" + name + "-" + version + ".tar.gz");
        writeAscii(
                output,
                "Content-Disposition: form-data; name=\"artifact\"; filename=\""
                        + filename
                        + "\"\r\n"
        );
        writeAscii(output, "Content-Type: application/octet-stream\r\n\r\n");
        output.write(artifact);
        writeAscii(output, "\r\n--" + boundary + "--\r\n");
        return output.toByteArray();
    }

    private static void writeAscii(ByteArrayOutputStream output, String text) throws IOException {
        output.write(text.getBytes(StandardCharsets.US_ASCII));
    }

    private static String safeFilename(String value) {
        return value.replaceAll("[^A-Za-z0-9._-]", "_");
    }

    private static byte[] readBounded(InputStream input, int limit, boolean failOnOverflow)
            throws IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream(Math.min(limit, 8192));
        byte[] buffer = new byte[8192];
        int remaining = limit;
        while (remaining > 0) {
            int read = input.read(buffer, 0, Math.min(buffer.length, remaining));
            if (read < 0) {
                return output.toByteArray();
            }
            output.write(buffer, 0, read);
            remaining -= read;
        }
        if (input.read() != -1 && failOnOverflow) {
            throw new ResponseTooLargeException(limit);
        }
        return output.toByteArray();
    }

    private static String sha256Hex(byte[] value) {
        try {
            return HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(value));
        } catch (NoSuchAlgorithmException impossible) {
            throw new IllegalStateException("SHA-256 is unavailable", impossible);
        }
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record PackageSummary(
            String org,
            String name,
            String description,
            String latest
    ) { }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record PackageMetadata(
            String org,
            String name,
            String description,
            String vcs,
            @JsonProperty("repo_url") String repoUrl,
            String latest,
            List<String> versions,
            @JsonProperty("version_scheme") String versionScheme
    ) { }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record VersionMetadata(
            String org,
            String name,
            String version,
            String sha256,
            long size,
            String format,
            @JsonProperty("vcs_tag") String vcsTag,
            @JsonProperty("vcs_commit") String vcsCommit,
            @JsonProperty("download_url") String downloadUrl,
            @JsonProperty("published_at") String publishedAt,
            boolean yanked
    ) { }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record SearchResponse(String query, List<PackageSummary> items) { }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record ClaimOrgResponse(String slug, boolean created) { }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record PublishResponse(String org, String name, String version, String sha256) { }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record YankResponse(String org, String name, String version, boolean yanked) { }

    public static final class ValidationException extends IllegalArgumentException {
        ValidationException(String message) {
            super(message);
        }

        ValidationException(String message, Throwable cause) {
            super(message, cause);
        }
    }

    public static final class MissingTokenException extends IllegalStateException {
        MissingTokenException() {
            super("authenticated registry operation requires a bearer token");
        }
    }

    public static final class ApiException extends IOException {
        private final int status;
        private final String code;
        private final String responseBody;

        ApiException(int status, String code, String responseBody) {
            super("registry error " + status + ": " + code);
            this.status = status;
            this.code = code;
            this.responseBody = responseBody;
        }

        public int status() {
            return status;
        }

        public String code() {
            return code;
        }

        public String responseBody() {
            return responseBody;
        }
    }

    public static class ResponseTooLargeException extends IOException {
        private final int limit;

        ResponseTooLargeException(int limit) {
            super("registry response exceeded " + limit + " bytes");
            this.limit = limit;
        }

        public int limit() {
            return limit;
        }
    }

    public static final class ArtifactTooLargeException extends ResponseTooLargeException {
        ArtifactTooLargeException(int limit) {
            super(limit);
        }
    }

    public static final class Sha256MismatchException extends IOException {
        private final String expected;
        private final String actual;

        Sha256MismatchException(String expected, String actual) {
            super("artifact SHA-256 mismatch");
            this.expected = expected;
            this.actual = actual;
        }

        public String expected() {
            return expected;
        }

        public String actual() {
            return actual;
        }
    }

    public static final class ProtocolException extends IOException {
        ProtocolException(String message, Throwable cause) {
            super(message, cause);
        }
    }
}

File: clients/kotlin/src/main/kotlin/tech/zpkg/client/ZedClient.kt
package tech.zpkg.client

import java.io.ByteArrayOutputStream
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.security.MessageDigest
import java.time.Duration
import java.util.UUID

class RegistryException(
    val status: Int,
    val code: String,
    val registryMessage: String,
) : RuntimeException("registry error $status: $code")

class ZedClient(
    registryUrl: String = "https://registry.zpkg.tech",
    private val token: String? = null,
    private val timeout: Duration = Duration.ofSeconds(30),
    private val http: HttpClient = HttpClient.newBuilder()
        .followRedirects(HttpClient.Redirect.NEVER)
        .connectTimeout(Duration.ofSeconds(10))
        .build(),
) {
    private val base: URI = normalizeBase(registryUrl)

    fun packageJson(org: String, name: String): String =
        json("GET", "/v1/packages/${segment(org)}/${segment(name)}")

    fun versionJson(org: String, name: String, version: String): String =
        json("GET", "/v1/packages/${segment(org)}/${segment(name)}/versions/${segment(version)}")

    fun searchJson(query: String): String =
        json("GET", "/v1/search?q=${URLEncoder.encode(query, StandardCharsets.UTF_8).replace("+", "%20")}")

    fun claimOrgJson(org: String): String =
        json("POST", "/v1/orgs", "{\"org\":${jsonString(org)}}", auth = true)

    fun setYankedJson(
        org: String,
        name: String,
        version: String,
        yanked: Boolean,
        reason: String? = null,
    ): String {
        val reasonField = reason?.let { ",\"reason\":${jsonString(it)}" } ?: ""
        return json(
            "POST",
            "/v1/packages/${segment(org)}/${segment(name)}/versions/${segment(version)}/yank",
            "{\"yanked\":$yanked$reasonField}",
            auth = true,
        )
    }

    fun yankJson(org: String, name: String, version: String, reason: String? = null): String =
        setYankedJson(org, name, version, true, reason)

    fun restoreJson(org: String, name: String, version: String): String =
        setYankedJson(org, name, version, false)

    fun publishJson(
        org: String,
        name: String,
        version: String,
        artifact: Path,
        metadataJson: String,
    ): String {
        require(Files.isRegularFile(artifact)) { "artifact must be a regular file" }
        val boundary = "zed-${UUID.randomUUID()}"
        val payload = multipart(boundary, metadataJson, artifact.fileName.toString(), Files.readAllBytes(artifact))
        return send(
            "PUT",
            "/v1/packages/${segment(org)}/${segment(name)}/versions/${segment(version)}",
            payload,
            "multipart/form-data; boundary=$boundary",
            auth = true,
            limit = MAX_JSON_BYTES,
        ).toString(StandardCharsets.UTF_8)
    }

    fun downloadArtifact(sha256: String, destination: Path? = null): ByteArray {
        val expected = sha256.lowercase()
        require(HEX_64.matches(expected)) { "sha256 must be 64 hexadecimal characters" }
        val bytes = send("GET", "/v1/artifacts/$expected", null, null, auth = false, limit = MAX_ARTIFACT_BYTES)
        val actual = MessageDigest.getInstance("SHA-256")
            .digest(bytes)
            .joinToString("") { "%02x".format(it) }
        if (actual != expected) {
            throw RegistryException(0, "digest_mismatch", "expected $expected, got $actual")
        }
        destination?.let { Files.write(it, bytes) }
        return bytes
    }

    private fun json(method: String, path: String, body: String? = null, auth: Boolean = false): String =
        send(
            method,
            path,
            body?.toByteArray(StandardCharsets.UTF_8),
            body?.let { "application/json" },
            auth,
            MAX_JSON_BYTES,
        ).toString(StandardCharsets.UTF_8)

    private fun send(
        method: String,
        path: String,
        body: ByteArray?,
        contentType: String?,
        auth: Boolean,
        limit: Int,
    ): ByteArray {
        val builder = HttpRequest.newBuilder(base.resolve(base.path.trimEnd('/') + path))
            .timeout(timeout)
            .header("Accept", "application/json")
        if (contentType != null) builder.header("Content-Type", contentType)
        if (auth && !token.isNullOrBlank()) builder.header("Authorization", "Bearer $token")
        builder.method(method, body?.let { HttpRequest.BodyPublishers.ofByteArray(it) } ?: HttpRequest.BodyPublishers.noBody())
        val response = http.send(builder.build(), HttpResponse.BodyHandlers.ofByteArray())
        if (response.body().size > limit) {
            throw RegistryException(0, "response_too_large", "response exceeded $limit bytes")
        }
        if (response.statusCode() !in 200..299) {
            val message = response.body().toString(StandardCharsets.UTF_8).take(16_384)
            throw RegistryException(response.statusCode(), "http_error", message)
        }
        return response.body()
    }

    private fun multipart(boundary: String, metadata: String, filename: String, artifact: ByteArray): ByteArray {
        val out = ByteArrayOutputStream()
        fun text(value: String) = out.write(value.toByteArray(StandardCharsets.UTF_8))
        val safeFilename = filename.replace(Regex("[\\r\\n\\\"]"), "_")
        text("--$boundary\r\nContent-Disposition: form-data; name=\"meta\"\r\nContent-Type: application/json\r\n\r\n")
        text(metadata)
        text("\r\n--$boundary\r\nContent-Disposition: form-data; name=\"artifact\"; filename=\"$safeFilename\"\r\nContent-Type: application/octet-stream\r\n\r\n")
        out.write(artifact)
        text("\r\n--$boundary--\r\n")
        return out.toByteArray()
    }

    private fun segment(value: String): String {
        require(value.isNotBlank() && value != "." && value != "..") { "path segment must be nonblank" }
        require(value.toByteArray(StandardCharsets.UTF_8).size <= 256) { "path segment is too long" }
        require(value.none { it.code < 0x20 || it.code == 0x7f }) { "path segment contains a control character" }
        return URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20")
    }

    private fun jsonString(value: String): String = buildString {
        append('"')
        value.forEach { c ->
            when (c) {
                '"' -> append("\\\"")
                '\\' -> append("\\\\")
                '\b' -> append("\\b")
                '\u000c' -> append("\\f")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                '\t' -> append("\\t")
                else -> if (c.code < 0x20) append("\\u%04x".format(c.code)) else append(c)
            }
        }
        append('"')
    }

    companion object {
        private const val MAX_JSON_BYTES = 16 * 1024 * 1024
        private const val MAX_ARTIFACT_BYTES = 100 * 1024 * 1024
        private val HEX_64 = Regex("^[0-9a-f]{64}$")

        private fun normalizeBase(raw: String): URI {
            val uri = URI(raw.trim())
            require(uri.scheme in setOf("http", "https") && uri.host != null) { "registryUrl must be absolute HTTP(S)" }
            require(uri.userInfo == null && uri.query == null && uri.fragment == null) { "registryUrl must not contain credentials, query, or fragment" }
            val cleanPath = uri.path.trimEnd('/')
            return URI(uri.scheme, null, uri.host, uri.port, cleanPath, null, null)
        }
    }
}
